src/main/java/fi/aalto/cs/apluscourses/presentation/base/SelectableNodeViewModel.java

```diff
@@ -31,18 +32,18 @@ public SelectableNodeViewModel(@NotNull T model,
    * @param filter A filter.
    * @return True, if the filter applies to this node or one of its descendants, otherwise false.
    */
-  public boolean applyFilter(Filter filter) {
-    if (Thread.currentThread().isInterrupted()) {
-      return true;
+  public Optional<Boolean> applyFilter(Filter filter) {
+    Optional<Boolean> myResult = filter.apply(this);
+    Optional<Boolean> result = myResult;
+    for (SelectableNodeViewModel<?> child : children) {
+      if (Thread.currentThread().isInterrupted()) {
+        return Optional.empty();
+      }
+      result = OptionalBooleanLogic.or(result, child.applyFilter(filter));
```

Comment:
I personally liked the non-abstracted version more :smile: It's not completely obvious how `OptionalBooleanLogic` deals with empty optionals, so now this adds an additional indirection which has to be checked every time the code is read

Reply:
Removed `OptionalBooleanLogic` I also changed it to skip child nodes if the result is false (node is hidden). Furthermore, I made it throw `InterruptedException` if interrupted to follow the convention.